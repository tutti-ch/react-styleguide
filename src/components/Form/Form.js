import React from "react";
import Input from "./_Input";
import InputCheckbox from "./_InputCheckbox";
import InputRadio from "./_InputRadio";
import Textarea from "./_Textarea";
import Slider from "./_Slider";
import Table from "../Table";
import InputRadioGroup from "./_InputRadioGroup";
import Select from "./_Select";
import Form from "./_Form";
import Button from "../Button";

/**
 * =====================================================================
 * This is an example form. Scroll Down to see the exported component. =
 * =====================================================================
 */
class ExampleForm extends React.Component {
  constructor(props) {
    super(props);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.times = 0;
    this.state = {
      values: null
    };
  }

  // istanbul ignore next
  handleSubmit(values) {
    if (this.times++ === 0) {
      return Promise.reject({
        name: "This is wrong",
        radios: "Error message",
        _error: "This is a generic error. Just pass in an _error key."
      });
    }

    this.setState({ values });
    return new Promise(resolve => {
      setTimeout(resolve, 1000);
    });
  }

  render() {
    return (
      <Form handleSubmit={this.handleSubmit}>
        <div style={{ marginBottom: "1rem" }}>
          If you check the source code, this is a nested form.
        </div>
        <div style={{ marginBottom: "1rem" }}>
          <Form.Select
            name="framework"
            placeholder="Select your favorite framework"
            options={[
              { text: "React", value: "react" },
              { text: "Vue.js", value: "vue" }
            ]}
          />
        </div>
        <div>
          <Form.Input type="text" name="name" value="This is a default value" />
          <Form.RadioGroup name="radios" label="Radio Ga Ga" inline>
            <Form.Radio label="Radio A" value="a" />
            <Form.Radio
              label="Radio B"
              value="b"
              style={{ marginLeft: "10px" }}
            />
            <Form.Radio
              label="Radio C"
              value="c"
              style={{ marginLeft: "10px" }}
            />
          </Form.RadioGroup>
        </div>
        <div>
          <div>
            <span style={{ marginRight: "10px", display: "inline-block" }}>
              Checkbox group:
            </span>
            <Form.Checkbox
              name="cb"
              value="Hey0"
              checked
              inline
              label="Hey 0"
            />
            <Form.Checkbox
              name="cb"
              value="Hey1"
              inline
              label="Hey 1"
              style={{ marginLeft: "10px" }}
            />
            <Form.Checkbox
              name="cb"
              value="Hey2"
              checked
              inline
              label="Hey 2"
              style={{ marginLeft: "10px" }}
            />
          </div>
          <div>
            <span style={{ marginRight: "10px", display: "inline-block" }}>
              Single checkboxes:
            </span>
            <Form.Checkbox
              name="cb-single-1"
              value="CB 1"
              checked
              inline
              label="Single cb 1"
            />
            <Form.Checkbox
              name="cb-single-2"
              value="CB 2"
              checked
              inline
              label="Single cb 2"
            />
          </div>
        </div>
        <div style={{ margin: "1.5rem 1rem 2.5rem 0" }}>
          <Form.Toggle value="Trans" name="gender">
            <Form.Toggle.Option value="Man">Man</Form.Toggle.Option>
            <Form.Toggle.Option value="Female">Female</Form.Toggle.Option>
            <Form.Toggle.Option value="Trans">Trans</Form.Toggle.Option>
          </Form.Toggle>
        </div>
        <div style={{ margin: "1.5rem 1rem 1rem 1rem" }}>
          <Form.Slider
            name={["minPrice", "maxPrice"]}
            label="Price range"
            values={[150, 230]}
            min={100}
            max={250}
            step={10}
            multiple
          />
          <Form.GenericError />
          <pre
            style={
              this.state.values
                ? {
                    backgroundColor: "#fafafa",
                    border: "1px solid #c5c5c5",
                    padding: "0.25rem"
                  }
                : {}
            }
          >
            <code>
              {this.state.values &&
                "Values: " + JSON.stringify(this.state.values, null, 2)}
            </code>
          </pre>
        </div>
        <div>
          <Button type="submit" level={Button.LEVEL_SECONDARY}>
            Submit
          </Button>
        </div>
      </Form>
    );
  }
}

/**
 * This component is only used by the style guide.
 * It should not be used in production.
 */
export default () => (
  <Table>
    <Table.Body>
      <Table.Row>
        <Table.Cell>
          <h3>Inputs</h3>
        </Table.Cell>
        <Table.Cell colSpan={4}>
          <Input label="Block text input" error="With error message" />
          <Input label="Inline text input" inline />
          <Input
            label="Number"
            type="number"
            min="100"
            max="1000"
            step="50"
            value="150"
          />
          <Input label="Password" type="password" />
          <Input label="Email" type="email" />
          <Input label="Phone" type="phone" />
        </Table.Cell>
      </Table.Row>
      <Table.Row>
        <Table.Cell>
          <h3>Slider</h3>
        </Table.Cell>
        <Table.Cell colSpan={4}>
          <div style={{ marginTop: "2rem", marginBottom: "4rem" }}>
            <Slider
              label="Slider { step: 750, minDistance: 2000 }"
              min={1000}
              max={10000}
              values={[1300, 5780]}
              minDistance={2000}
              name={["minDist", "maxDist"]}
              prefix={["From ", "To "]}
              suffix=".-"
              step={750}
              multiple
            />
          </div>
          <div style={{ marginTop: "2rem", marginBottom: "4rem" }}>
            <Slider
              label="Slider { step: 1000, minDistance: 0 }"
              min={1000}
              max={10000}
              values={[1300, 5780]}
              minDistance={0}
              name={["minDist", "maxDist"]}
              prefix={["From ", "To "]}
              suffix=".-"
              step={1000}
              multiple
            />
          </div>
          <div style={{ marginTop: "2rem", marginBottom: "4rem" }}>
            <Slider
              label="Slider single"
              range={[
                { value: "1", label: "OMG! SAVAS <3" },
                { value: "2", label: "Wow! Savas is great" },
                { value: "3", label: "It is great, Thanks Savas " },
                {
                  value: "4",
                  label: "I like it very much, Savas is fantastic"
                },
                { value: "5", label: "I love it, Savas is awesome." }
              ]}
              values={["5"]}
            />
          </div>
          <div style={{ marginTop: "2rem", marginBottom: "4rem" }}>
            <Slider
              label="Slider { range: Array, minDistance: 1, step: 1 }"
              range={[
                { value: "1", label: "CHF 25" },
                { value: "2", label: "CHF 35" },
                { value: "3", label: "CHF 50" },
                { value: "4", label: "CHF 100" },
                { value: "5", label: "CHF 500" },
                { value: "6", label: "CHF 1'000" }
              ]}
              values={["2", "5"]}
              name={["minDist", "maxDist"]}
              minDistance={1}
              prefix={["From ", "To "]}
              step={1}
              multiple
            />
          </div>
          <div>
            <Slider
              label="Slider { range: Array, minDistance: 0, step: 1, extremes, mouseThreshold: 10 }"
              range={[
                { value: "1", label: "CHF 25" },
                { value: "2", label: "CHF 35" },
                { value: "3", label: "CHF 50" },
                { value: "4", label: "CHF 100" },
                { value: "5", label: "CHF 500" },
                { value: "6", label: "CHF 1'000" }
              ]}
              name={["minDist", "maxDist"]}
              values={[null, null]}
              minDistance={0}
              prefix={["From ", "To "]}
              mouseThreshold={10}
              step={1}
              extremes
              multiple
            />
          </div>
        </Table.Cell>
      </Table.Row>
      <Table.Row>
        <Table.Cell>
          <h3>Checkbox</h3>
        </Table.Cell>
        <Table.Cell colSpan={4}>
          <InputCheckbox value="red" label="Red" checked />
          <InputCheckbox value="blue" label="Blue" checked />
          <InputCheckbox value="green" label="Green" />
        </Table.Cell>
      </Table.Row>
      <Table.Row>
        <Table.Cell>
          <h3>Button Group</h3>
        </Table.Cell>
        <Table.Cell colSpan={4}>
          <InputRadioGroup>
            <InputRadio checked value={10} label="React.js" name="r1" />
            <InputRadio value={10} label="Vue.js" name="r1" />
            <InputRadio value={10} label="Angular.js" name="r1" />
          </InputRadioGroup>
        </Table.Cell>
      </Table.Row>
      <Table.Row>
        <Table.Cell>
          <h3>Textarea</h3>
        </Table.Cell>
        <Table.Cell colSpan={4}>
          <Textarea label="This is a textarea" />
        </Table.Cell>
      </Table.Row>
      <Table.Row>
        <Table.Cell>
          <h3>Toggle</h3>
        </Table.Cell>
        <Table.Cell colSpan={4}>
          <div>
            <Form.Toggle value="Trans" name="gender">
              <Form.Toggle.Option value="Man">Man</Form.Toggle.Option>
              <Form.Toggle.Option value="Female">Female</Form.Toggle.Option>
              <Form.Toggle.Option value="Trans">Trans</Form.Toggle.Option>
            </Form.Toggle>
          </div>
          <div style={{ marginTop: "2rem" }}>
            <Form.Toggle value="Toggle" name="type" disabled>
              <Form.Toggle.Option value="Man">Disabled</Form.Toggle.Option>
              <Form.Toggle.Option value="Toggle">Toggle</Form.Toggle.Option>
            </Form.Toggle>
          </div>
        </Table.Cell>
      </Table.Row>
      <Table.Row>
        <Table.Cell>Dropdown</Table.Cell>
        <Table.Cell colSpan={4}>
          <Select
            label="Block select input"
            placeholder="Favorite framework"
            options={[
              { value: "React.js", text: "React.js" },
              { value: "Vue.js", text: "Vue.js" },
              { value: "Angular.js", text: "Angular.js" },
              { value: "jQuery", text: "jQuery" },
              { value: "Knockout", text: "Knockout" },
              { value: "Backbone", text: "Backbone" },
              { value: "Ember", text: "Ember" },
              { value: "Bootstrap", text: "Bootstrap" },
              { value: "Other", text: "Other" }
            ]}
          />
          <Select
            label="Default Value for Select"
            placeholder="Favorite framework"
            selected="React.js"
            options={[
              { value: "React.js", text: "React.js" },
              { value: "Vue.js", text: "Vue.js" },
              { value: "Angular.js", text: "Angular.js" },
              { value: "jQuery", text: "jQuery" },
              { value: "Knockout", text: "Knockout" },
              { value: "Backbone", text: "Backbone" },
              { value: "Ember", text: "Ember" },
              { value: "Bootstrap", text: "Bootstrap" },
              { value: "Other", text: "Other" }
            ]}
          />
          <Select
            label="Multiple select input"
            multiple
            placeholder="Favorite framework"
            options={[
              { value: "React.js", text: "React.js" },
              { value: "Vue.js", text: "Vue.js" },
              { value: "Angular.js", text: "Angular.js" },
              { value: "jQuery", text: "jQuery" },
              { value: "Knockout", text: "Knockout" },
              { value: "Backbone", text: "Backbone" },
              { value: "Ember", text: "Ember" },
              { value: "Bootstrap", text: "Bootstrap" },
              { value: "Other", text: "Other" }
            ]}
          />
          <Select
            label="Inline select input"
            inline
            placeholder="Favorite framework"
            options={[
              { value: "React.js", text: "React.js" },
              { value: "Vue.js", text: "Vue.js" },
              { value: "Angular.js", text: "Angular.js" },
              { value: "jQuery", text: "jQuery" },
              { value: "Knockout", text: "Knockout" },
              { value: "Backbone", text: "Backbone" },
              { value: "Ember", text: "Ember" },
              { value: "Bootstrap", text: "Bootstrap" },
              { value: "Other", text: "Other" }
            ]}
          />
        </Table.Cell>
      </Table.Row>
      <Table.Row>
        <Table.Cell>Form</Table.Cell>
        <Table.Cell colSpan={4}>
          <ExampleForm />
        </Table.Cell>
      </Table.Row>
    </Table.Body>
  </Table>
);
