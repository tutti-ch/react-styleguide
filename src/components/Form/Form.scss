@import "../../styles/main";

// Form variables

$inputHeight: $size-xxxl;

@mixin placeholder {
  // IMPORTANT: Do not combine the following placeholder styles, as otherwise they will break.

  // Chrome/Opera/Safari
  ::-webkit-input-placeholder {
    @content;
  }

  // Firefox 19+
  :-moz-placeholder {
    @content;
  }

  // IE 10+
  ::-moz-placeholder {
    @content;
  }

  // Firefox 18-
  :-ms-input-placeholder {
    @content;
  }
}

@include placeholder {
  color: $greyscale-60;
}

@mixin focus {
  transition: $transition-m;

  &:focus {
    outline: none;
    border-color: $greyscale-00;
  }
}

@include shakeAnimation();

.has-error {
  animation: shake 0.82s cubic-bezier(0.36, 0.07, 0.19, 0.97) both;
  transform: translate3d(0, 0, 0);
  backface-visibility: hidden;
  perspective: 1000px;
}

.generic-error {
  background: $red-50;
  color: $greyscale-100;
  padding: $size-xxs;
  border-radius: $border-radius-s;
}

.wrapper {
  min-height: $size-xxxl;
  display: flex;
  align-items: center;
  position: relative;
  font-size: 1rem;

  .error {
    position: absolute;
    top: 100%;
    right: 0;
    margin-top: $size-xxs;
    background: $red-50;
    color: $greyscale-100;
    padding: $size-xxxs $size-xxs;
    border-radius: $size-xxxs;
    font-size: $size-xs;
    white-space: pre-wrap;
    max-width: 55%;
    @include z-index("overlay", "ground");

    &:before {
      display: block;
      content: "";
      position: absolute;
      right: $size-m;
      bottom: 100%;
      width: 0;
      height: 0;
      border-left: $size-xxs solid transparent;
      border-right: $size-xxs solid transparent;
      border-bottom: $size-xxs solid $red-50;
    }
  }

  &.inline {
    display: inline-flex;
  }

  & + .wrapper {
    margin-top: $size-xl;
  }

  & + label.wrapper {
    margin-top: $size-xl;
  }

  input[type="text"],
  input[type="password"],
  input[type="email"],
  input[type="phone"],
  input[type="number"],
  textarea {
    border: $border-default;
    background-color: $greyscale-100;
    border-radius: 2px;
    position: relative;
    box-sizing: border-box;
    width: 100%;
    @include focus();
    min-height: $size-xxxl;
  }

  input {
    line-height: $inputHeight;
    padding: 0 $size-s;
  }

  textarea {
    min-height: 8rem;
    resize: none;
    padding: $size-s;
  }

  textarea + .label {
    top: $size-s;
    transform: translateY(0);
  }

  input:focus + .label,
  textarea:focus + .label,
  .slider .label,
  &:not(.inline) .select + .label,
  &.has-value:not(.select-wrapper) .label {
    font-size: 0.85rem;
    top: -$size-xxxs;
    left: 0;
    transform: translateY(-100%);
  }

  input[type="checkbox"],
  input[type="radio"] {
    font: inherit;
    vertical-align: middle;
    background: $greyscale-100;
    border-radius: $size-xxxs / 2;
    box-sizing: border-box;
    border: $border-default;
    position: relative;
    margin-right: $size-m;
  }

  .cb-label {
    user-select: none;
    cursor: pointer;
  }

  .label {
    position: absolute;
    pointer-events: none;
    left: $size-s;
    top: 50%;
    color: $greyscale-60;
    transform: translateY(-50%);
    transition: $transition-s;
  }

  .slider .label,
  &.has-value .slider .label {
    top: -$size-m;
  }

  /**
  * STYLES FOR THE SLIDER
  */
  $slider-height: $size-xxs;

  .slider {
    position: relative;
    display: inline-block;
    width: 100%;
    height: $slider-height;
    background: $greyscale-50;
    border-radius: $size-xxxs / 2;
    border: 0 solid #010101;
    outline: none;
    user-select: none;

    .label-text {
      display: inline-block;
      margin-right: $size-xs;

      &:after {
        content: ":";
      }
    }

    .range-min {
      display: inline-block;
      min-width: 5.5rem;
    }

    .range-min + .range-max {
      margin-left: $size-xs;
    }

    $thumb-size: $size-xl;

    .thumbBg {
      display: inline-block;
      height: $slider-height;
      background: $blue-50;
      position: absolute;
      margin-left: $thumb-size;
    }

    .thumb {
      display: inline-block;
      border: $border-default;
      height: $thumb-size;
      width: $thumb-size;
      border-radius: 50%;
      background: #ffffff;
      outline: none;
      cursor: pointer;
      position: absolute;
      transition: width 0.1s ease-out, height 0.1s ease-out;
      top: 50%;
      transform: translateY(-50%);
      user-select: none;

      &.trans {
        transition: $transition-m;
      }

      &[data-value]:before {
        position: absolute;
        content: attr(data-value);
        left: -50%;
        right: -50%;
      }

      &:hover {
        border: 1px solid $greyscale-80;
      }
    }
  }

  /**
  * STYLES FOR THE SELECT
  */
  .select {
    position: relative;
    outline: none;
    user-select: none;
    width: 100%;

    // Styles for when the placeholder is active
    &.placeholder {
      position: relative;

      .ph {
        color: $greyscale-60;
      }
    }

    .ph {
      display: flex;
      flex-wrap: wrap;
      max-width: 100%;
      position: relative;
      flex: 1;

      .option {
        max-width: 100%;
        flex: 0 0 auto;
      }

      .option:not(:last-child):after {
        content: ", ";
      }

      // Styles for the arrow
      &:after {
        content: "";
        border-color: $greyscale-80 transparent transparent;
        border-style: solid;
        border-width: 0.4rem 0.4rem 0.2rem;
        display: inline-block;
        height: 0;
        width: 0;
        top: 50%;
        transition: all $transition-m;
        right: - $size-xs * 2;
        margin-top: 0.2rem;
        position: absolute;
        transform: translateY(-50%);
      }
    }

    .select-inner {
      max-width: 100%;
      border: $border-default;
      background: $greyscale-100;
      min-height: $size-xxxl;
      line-height: $size-xxxl;
      position: relative;
      text-align: left;
      padding: 0 $size-xs * 3 0 $size-xs;
      cursor: pointer;
      display: flex;
      align-items: center;

      &:focus {
        border-color: $greyscale-00;
        outline: none;
      }

      .options {
        background: $greyscale-100;
        position: absolute;
        margin-top: -2px;
        margin-bottom: 4rem; // When the options hit the bottom of page, this gives a bit of space
        top: 100%;
        right: -1px;
        left: -1px;
        opacity: 0;
        max-height: 0;
        visibility: hidden;
        overflow: auto;
        z-index: 1;
        @include mediumShadow();
        border: $border-default;
        user-select: none;
        @include z-index("overlay", "ground");

        // Keep the style separated for the placeholder option
        .option {
          padding: 0 $size-xs;
          border-bottom: 1px solid $greyscale-95;

          &:hover,
          &.highlighted {
            background-color: $blue-95;
          }

          &.selected {
            font-weight: bold;
          }
        }
      }

      .option {
        display: flex;
        align-items: center;
      }

      // This is the placeholder
      .option.placeholder {
        color: $greyscale-60;

        &.selected {
          font-weight: normal;
        }
      }
    }

    .text {
      @include textTruncate;
    }

    .icon {
      margin-right: $size-xs;
      width: $size-xl;
      height: $size-xl;

      @include ifInternetExplorer {
        width: 2.2rem;
        height: 2.2rem;
      }
    }

    // When the select box is open or is hover make the border black
    &.is-open .options,
    &.is-open .select-inner,
    &:not(.disabled):hover .select-inner {
      border-color: $greyscale-00;
    }

    &.is-open {
      .options {
        max-height: 320px;
        opacity: 1;
        visibility: visible;
      }

      &:not(.inline) {
        border-top-color: $greyscale-80;
      }
    }

    &:not(.disabled):hover .ph:after,
    &.is-open .ph:after {
      border-color: $greyscale-00 transparent transparent;
    }

    &.disabled {
      .select-inner {
        cursor: default;
        background: $greyscale-95;
        color: $greyscale-60;
      }

      .icon {
        opacity: 0.6;
      }
    }
  }

  /**
  * Styles for an inline select
  */
  &.inline.select-wrapper {
    align-items: center;
    line-height: 1;
    font-size: 1rem;

    &.hasLabel {
      flex-direction: row-reverse;
    }

    .select {
      flex: 0 0 auto;
      width: auto;
      max-width: 100%;
    }

    .label {
      flex: 1 1 auto;
      position: static;
      top: auto;
      left: auto;
      transform: translateY(0);
    }

    .select-inner {
      border: 1px solid transparent;
      display: inline-block;
      background-color: transparent;
      height: auto;

      .options .option {
        line-height: $size-xxxl;
        height: $size-xxxl;
      }

      &:not(.disabled):hover {
        border: 1px solid transparent;
      }
    }

    .options {
      left: auto;
      min-width: 250px;
      max-height: 320px;
    }
  }

  .icoClose,
  .icoCheck {
    display: none;
  }

  .select.multiple {
    .ph {

      // Styles for the arrow
      &:after {
        top: $size-xxxl / 2;
        transform: none;
        margin-top: 0;
      }
    }

    &:not(.placeholder) {
      .ph {
        height: 100%;
        line-height: 1rem;
        align-items: center;
        padding-top: $size-s;

        .option {
          background-color: $blue-50;
          color: $greyscale-100;
          border-radius: 3px;
          padding: $size-xxxs $size-xxs;
          margin: 0 $size-s $size-s 0;
          transition: all $transition-m;

          &:hover {
            background-color: $blue-60;
          }

          &:after {
            display: none;
          }
        }

        .icoClose {
          display: block;
          margin: 0 0 0 $size-xxs;
          transition: all $transition-s;

          &:hover {
            color: $blue-20;
          }

          &:before {
            margin: 0;
            font-size: $size-xs;
          }
        }

        .icoCheck {
          display: none;
        }
      }
    }

    .options .option.selected {
      position: relative;

      .icoCheck {
        content: "";
        display: inline-block;
        position: absolute;
        right: $size-xs;
        color: $blue-50;
        margin: 0;
      }
    }
  }
}

.radioLabel {
  pointer-events: none;
  color: $greyscale-60;
  font-size: 0.85rem;
  margin-right: $size-m;
}

.toggleWrapper {
  border: $border-default;
  border-radius: $border-radius-s;

  &.disabled {
    pointer-events: none;

    .toggleOption {
      background-color: transparent;
    }

    .toggleSelected {
      background-color: $greyscale-60;
    }
  }

  .toggleOption {
    display: inline-block;
    padding: $size-m;
    cursor: pointer;
    border-right: $border-default;
    background-color: $greyscale-100;

    &:last-child {
      border-right: none;
    }
  }

  .toggleSelected {
    background: $blue-50;
    color: $greyscale-100;
  }
}
