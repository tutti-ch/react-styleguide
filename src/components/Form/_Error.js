import React, { Component } from "react";
import PropTypes from "prop-types";
import classes from "./Form.scss";

export default class Error extends Component {
  static contextTypes = {
    genericErrorHandler: PropTypes.func
  };

  constructor(props, context) {
    super(props);

    this.setError = this.setError.bind(this);
    this.state = {
      error: null
    };

    if (typeof context.genericErrorHandler === "function") {
      context.genericErrorHandler(this);
    }
  }

  setError(message) {
    this.setState({ error: message });
  }

  render() {
    const { error } = this.state;

    if (!error) {
      return null;
    }

    return <div className={classes.genericError}>{error}</div>;
  }
}
