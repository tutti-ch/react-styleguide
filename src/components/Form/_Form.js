import React, { Component } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import Input from "./_Input";
import Textarea from "./_Textarea";
import Radio from "./_InputRadio";
import RadioGroup from "./_InputRadioGroup";
import Checkbox from "./_InputCheckbox";
import Select from "./_Select";
import Stepper from "./_Stepper";
// import Slider from "./_Slider"; removed slider as we don't use it at the moment
import Toggle from "./_Toggle";
import Error from "./_Error";
import { filterProps } from "../../helpers/functions";
import classes from "./Form.scss";

export default class Form extends Component {
  static Input = Input;
  static Radio = Radio;
  static RadioGroup = RadioGroup;
  static Checkbox = Checkbox;
  static Select = Select;
  // static Slider = Slider;
  static Textarea = Textarea;
  static GenericError = Error;
  static Toggle = Toggle;
  static Stepper = Stepper;

  static propTypes = {
    className: PropTypes.oneOfType([
      PropTypes.object,
      PropTypes.array,
      PropTypes.string
    ]),
    children: PropTypes.node,
    handleSubmit: PropTypes.func // Function to handle submit.
  };

  static childContextTypes = {
    register: PropTypes.func,
    onSubmit: PropTypes.func,
    genericErrorHandler: PropTypes.func
  };

  /**
   * Assign key value pair to the first object. In case a property
   * already exists, this function will aggregate the results.
   *
   * @example
   * arguments: { a: "b" }, { a: "c", d: 1 }
   * returns: { a: ["b", "c"], d: 1 }
   *
   * @param values
   * @param keyValuePair
   */
  static assignKeyPair(values, keyValuePair) {
    Object.keys(keyValuePair).forEach(k => {
      if (values[k]) {
        // Cast to an array ({ a: "b" } becomes { a: ["b"] })
        if (!Array.isArray(values[k])) {
          values[k] = [values[k]];
        }

        values[k].push(keyValuePair[k]);
      } else {
        values[k] = keyValuePair[k];
      }
    });

    return values;
  }

  inputs = [];
  listeners = [];
  submitting = false;

  getChildContext() {
    return {
      /**
       * Callback function to let the child component register itself on the form.
       * Once registered, the Form component will require the child to have two methods:
       * - handleError
       * - getKeyValue
       *
       * @param {object|Array} props
       * @param {string} props.name
       * @param {*} props.value
       * @param {*} props.values Some components (Slider) uses the plural name
       * @param {object} self The instance.
       */
      register: (props, self) => {
        const { name, value } = props;
        this.inputs = this.inputs || [];
        this.inputs.push({
          defaultValue: value || props.values, // Values is used by slider
          error: null,
          name,
          self
        });
      },

      /**
       * Register a listener for the submit event.
       *
       * @param listener
       */
      onSubmit: listener => {
        this.listeners = this.listeners || [];
        this.listeners.push(listener);
      },

      /**
       * This is the generic error handler. In case the handleSubmit reject
       * passes an _error key, it will be displayed by this component.
       *
       * @param self
       */
      genericErrorHandler: self => {
        this.genericErrorHandler = self;
      }
    };
  }

  handleSubmit = e => {
    if (typeof this.props.handleSubmit !== "function") {
      return false;
    }

    e.preventDefault();

    // Prevent re-submitting form until promise is resolved/rejected.
    if (this.submitting) return false;
    this.submitting = true;

    let values = {};

    // Collect all values from the registered inputs.
    this.inputs.forEach(input => {
      // Reset error state
      if (input.error) {
        input.error = null;
        input.self.handleError(null);
      }

      const { self } = input;
      values = Form.assignKeyPair(values, self.getKeyValue());
    });

    // Make sure to reset the generic error state
    if (this.genericError && this.genericErrorHandler) {
      this.genericErrorHandler.setError(null);
    }

    // Notify listeners and tell them that we are submitting.
    this.listeners.forEach(l => l({ values, loading: true }));

    // Execute the handleSubmit function and wait for a response/reject.
    return this.props
      .handleSubmit(values)
      .catch(errors => {
        if (errors !== null && typeof errors === "object") {
          Object.keys(errors).forEach(key => {
            for (let i = 0; i < this.inputs.length; i++) {
              const input = this.inputs[i];

              if (input.name === key) {
                input.self.handleError(errors[key]);
                input.error = errors[key];
                break;
              }
            }
          });

          if (errors._error && this.genericErrorHandler) {
            this.genericErrorHandler.setError(errors._error);
          }
        }

        // So that we display the shake.
        this.genericError = errors ? !!errors._error : false;
        this.forceUpdate();
      })
      .then(() => {
        this.listeners.forEach(l => l({ values, loading: false }));
        this.submitting = false;
      });
  };

  /**
   * Reset listeners.
   */
  componentWillUnmount() {
    this.listeners = [];
  }

  render() {
    const props = filterProps(Form.propTypes, this.props);
    const formClasses = classNames(this.props.className, {
      [classes.hasError]:
        this.inputs.filter(i => i.error).length > 0 || this.genericError
    });

    return (
      <form {...props} onSubmit={this.handleSubmit} className={formClasses}>
        {this.props.children}
      </form>
    );
  }
}
