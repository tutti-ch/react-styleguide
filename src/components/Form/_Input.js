import React, { Component } from "react";
import PropTypes from "prop-types";
import { filterProps } from "../../helpers/functions";
import WithWrapper from "./_WithWrapper";

// istanbul ignore next
const MutationObserver =
  typeof window !== "undefined"
    ? window.MutationObserver ||
      window.WebKitMutationObserver ||
      window.MozMutationObserver
    : undefined;

export class Input extends Component {
  static defaultProps = {
    value: "",
    type: "text"
  };

  static propTypes = {
    /**
     * The initial value.
     */
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
      PropTypes.bool
    ]),

    /**
     * The input type.
     */
    type: PropTypes.string,

    /**
     * The input name.
     */
    name: PropTypes.string
  };

  constructor(props) {
    super(props);

    this.state = {
      value: props.value
    };

    this.notify = this.notify.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.handleBlur = this.handleBlur.bind(this);
  }

  // istanbul ignore next
  componentDidMount() {
    // This code fixes a bug which happens when a user saves information
    // such as login information, and the browser fills the input automatically.
    if (!this.props.value && MutationObserver) {
      this.observer = new MutationObserver(mutations => {
        mutations.forEach(mutation => {
          const target = mutation.target;

          if (
            target.defaultValue &&
            mutation.type === "attributes" &&
            mutation.attributeName === "defaultValue"
          ) {
            this.changed = true;
            this.setState({ value: target.defaultValue });
            this.notify();
            this.observer = this.observer.disconnect();
            target.defaultValue = ""; // Reset so that react does not complain about controlled inputs
          }
        });
      });

      this.observer.observe(this.refs.input, {
        attributes: true
      });
    }
  }

  // istanbul ignore next
  componentWillUnmount() {
    if (this.observer) {
      this.observer.disconnect();
    }
  }

  /**
   * Sync the value.
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.value !== this.props.value) {
      this.changed = true;
      this.setState({ value: nextProps.value || "" }, this.notify);
    }
  }

  /**
   * Notify the parent of the change.
   */
  notify() {
    if (this.changed === true && typeof this.props.onChange === "function") {
      const { name, value } = this.props;
      const newValue = this.state.value;
      this.props.onChange(newValue, {
        name,
        initialValue: value,
        formValue: newValue
      });
    }

    // Reset the value
    this.changed = false;
  }

  /**
   * Notify the parent when blur occurs.
   *
   * @param event
   */
  handleBlur(event) {
    this.notify();

    // istanbul ignore else
    if (typeof this.props.onBlur === "function") {
      this.props.onBlur(event);
    }
  }

  /**
   * Notify the parent when user presses enter.
   *
   * @param event
   */
  handleKeyDown(event) {
    if (event.key === "Enter") {
      this.notify();
    }

    if (typeof this.props.onKeyDown === "function") {
      this.props.onKeyDown(event);
    }
  }

  /**
   * Handle the change event.
   */
  handleChange(event) {
    this.setState({ value: event.target.value }, () => {
      // Next time when on blur or key up occurs,
      // they will know that the value changed.
      this.changed = true;
    });
  }

  render() {
    const { value } = this.state;
    const { type, name } = this.props;
    const props = filterProps(Input.propTypes, this.props);

    return (
      <input
        {...props}
        name={name}
        type={type}
        value={value}
        onBlur={this.handleBlur}
        onKeyDown={this.handleKeyDown}
        onChange={this.handleChange}
        ref="input"
      />
    );
  }
}

export default WithWrapper(Input);
