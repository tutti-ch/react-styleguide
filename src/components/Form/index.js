export { default } from "./_Form";
