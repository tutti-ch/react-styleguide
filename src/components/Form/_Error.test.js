/* global describe, test, jest, expect */
import React from "react";
import Form from "./index";
import { mount } from "enzyme";

describe("(Component) Form.GenericError", () => {
  test("should match the snapshots", done => {
    const handleSubmit = jest
      .fn()
      .mockReturnValue(Promise.reject({ _error: "Hey" }));
    const comp = mount(
      <Form handleSubmit={handleSubmit}>
        <Form.GenericError />
      </Form>
    );
    expect(comp).toMatchSnapshot();

    comp
      .instance()
      .handleSubmit({ preventDefault: jest.fn() })
      .then(() => {
        comp.update();
        expect(comp).toMatchSnapshot();
        done();
      });
  });
});
