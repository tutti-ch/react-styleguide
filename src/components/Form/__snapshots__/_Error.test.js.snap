// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) Form.GenericError should match the snapshots 1`] = `
<Form
  handleSubmit={[Function]}
>
  <form
    className=""
    onSubmit={[Function]}
  >
    <Error />
  </form>
</Form>
`;

exports[`(Component) Form.GenericError should match the snapshots 2`] = `
<Form
  handleSubmit={[Function]}
>
  <form
    className="undefined"
    onSubmit={[Function]}
  >
    <Error>
      <div>
        Hey
      </div>
    </Error>
  </form>
</Form>
`;
